cni/pkg/nodeagent/ztunnelserver.go

```diff
@@ -148,8 +149,16 @@ func (z *ztunnelServer) Close() error {
 func (z *ztunnelServer) Run(ctx context.Context) {
 	context.AfterFunc(ctx, func() { _ = z.Close() })
 
+	// Allow at most 5 requests per second. This is still a ridiculous amount; at most we should have 2 ztunnels on our node,
```

Comment:
We already have a safeguard that was merged here: https://github.com/istio/ztunnel/pull/1180

which does the same thing?

Can you reproduce what you're seeing with that in place?

If we are worried about `istio-cni` crashing machines during development for the older 1.22 builds that do not have that in place, it would probably be best to put a max memory consumption limit on it and let k8s handle it - which we should probably do anyway.

Reply:
Just like with keepalives, putting it on both sides is defense in depth. There is zero reason to write code that generates an unbounded amount of resources consumption, especially when interacting with another component. 

keep in mind ztunnel is pluggable as well. Just because we may write a nice ztunnel, doesn't mean someone else will

Reply:
I get that, and am not opposed to this, but do I want to make sure we're fixing the right thing once for all, rather than solving a local itch by throwing some code at it across several components and leaving it there.


- Wanna make sure we are not masking bugs on the ztunnel side with this.
- This indicates to me we should set memlimits on `istio-cni` - in the spirit of `defense in depth` I would prefer our system node critical components to be unable to crash a cluster with OOM even if they _do_ have a memory leak.
- This still isn't a great solution and doesn't really prevent an OOM, just slows it down a bit.

I would rather put a hard cap on retries, and mark the `istio-cni` unhealthy or panic if we hit it. We don't have a protocol mismatch in `master` (unless there's a bug here or in `ztunnel` I'm missing), so a mismatch can only be fatal + nonrecoverable here anyway.

We can (and should) revisit that with a proper sort of protocol negotiation based on the `ZDShello` version if we _do_ rev/break the protocol, but we haven't yet.

Reply:
This is, IMO, a safer solution than all of those. Maybe we want to do all of them, but they do come with some risks:
* Panic after some limit has complex logic around when to reset
* OOM limit is risky; if its too small we can accidentally hit it which can do more harm than good (recall the issue where we buffered the binaries in memory for instance... it was like 150mb. A lot, but not enough I want it OOM killed).

I am not against doing those either, but this is a ~zero risk fix that seems like a no-brainer to me.

Reply:
My main concern is that since it mostly hides the problem rather than resolves it, the risk enters from being less likely to catch real issues in testing when we encounter them, due to this turning a fast crash into a slower one.

But I agree that's probably paranoia in this case.